source/common/common/regex.cc

```diff
@@ -42,17 +44,50 @@ class CompiledStdMatcher : public CompiledMatcher {
 class CompiledGoogleReMatcher : public CompiledMatcher {
 public:
   CompiledGoogleReMatcher(const envoy::type::matcher::v3::RegexMatcher& config)
-      : regex_(config.regex(), re2::RE2::Quiet) {
+      : regex_(config.regex(), re2::RE2::Quiet), runtime_(Runtime::LoaderSingleton::getExisting()) {
     if (!regex_.ok()) {
       throw EnvoyException(regex_.error());
     }
 
-    const uint32_t max_program_size =
-        PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.google_re2(), max_program_size, 100);
-    if (static_cast<uint32_t>(regex_.ProgramSize()) > max_program_size) {
-      throw EnvoyException(fmt::format("regex '{}' RE2 program size of {} > max program size of "
-                                       "{}. Increase configured max program size if necessary.",
-                                       config.regex(), regex_.ProgramSize(), max_program_size));
+    const uint32_t regex_program_size = static_cast<uint32_t>(regex_.ProgramSize());
+
+    // Check if the deprecated field max_program_size is set first, and follow the old logic if so.
+    if (config.google_re2().has_max_program_size()) {
+      const uint32_t max_program_size =
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.google_re2(), max_program_size, 100);
+      if (static_cast<uint32_t>(regex_program_size > max_program_size)) {
+        throw EnvoyException(fmt::format("regex '{}' RE2 program size of {} > max program size of "
+                                         "{}. Increase configured max program size if necessary.",
+                                         config.regex(), regex_program_size, max_program_size));
+      }
+      return;
+    }
+
+    if (runtime_) {
+      Stats::Store& stats_store = runtime_->getStore();
+      Stats::StatNameManagedStorage program_size_stat_name("regex.program_size",
+                                                           stats_store.symbolTable());
+      Stats::Histogram& program_size_stat = stats_store.histogramFromStatName(
+          program_size_stat_name.statName(), Stats::Histogram::Unit::Unspecified);
+      program_size_stat.recordValue(regex_program_size);
```

Comment:
I don't feel super strongly about it, but I would probably skip the histogram here. However, in the warn case I would definitely have a stat for each warn so operators can monitor this.

Either way the stat(s) need to be documented and we need a release note for this change.

Reply:
I see some value in this data for monitoring what kind of program sizes are being configured and probably tune max_program_size or tune their regexes?

Reply:
Sure if you think there is utility SGTM. Let's add the counter stat though for warnings and make sure they all get documented.

Reply:
sure. Big +1 counter stat for warnings + docs